src/protocols/http/hpr_roaming_protocol.erl

```diff
@@ -0,0 +1,541 @@
+    GWTime.
+
+-spec response_stream(#packet{}) -> hpr_router_stream_manager:gateway_stream().
+response_stream(#packet{response_stream = ResponseStream}) ->
+    ResponseStream.
+
+-spec make_uplink_payload(
+    NetID :: netid_num(),
+    Uplinks :: list(packet()),
+    TransactionID :: integer(),
+    ProtocolVersion :: pv_1_0 | pv_1_1,
+    DedupWindowSize :: non_neg_integer(),
+    Destination :: binary(),
+    FlowType :: sync | async,
+    RoutingInfo :: hpr_routing:routing_info()
+) -> prstart_req().
+make_uplink_payload(
+    NetID,
+    Uplinks,
+    TransactionID,
+    ProtocolVersion,
+    DedupWindowSize,
+    Destination,
+    FlowType,
+    RoutingInfo
+) ->
+    #packet{
+        packet_up = PacketUp,
+        gateway_time = GatewayTime,
+        response_stream = ResponseStream
+    } = select_best(Uplinks),
+    Payload = hpr_packet_up:payload(PacketUp),
+    PacketTime = hpr_packet_up:timestamp(PacketUp),
+
+    Region = hpr_packet_up:region(PacketUp),
+    DataRate = hpr_packet_up:datarate(PacketUp),
+    Frequency = hpr_packet_up:frequency_mhz(PacketUp),
+
+    {RoutingKey, RoutingValue} =
+        case RoutingInfo of
+            {devaddr, DevAddr} -> {'DevAddr', encode_devaddr(DevAddr)};
+            {eui, DevEUI, _AppEUI} -> {'DevEUI', encode_deveui(DevEUI)}
+        end,
+
+    Token = make_uplink_token(TransactionID, Region, PacketTime, Destination, FlowType),
+
+    ok = hpr_roaming_downlink:insert_handler(TransactionID, ResponseStream),
+
+    VersionBase =
+        case ProtocolVersion of
+            pv_1_0 ->
+                #{'ProtocolVersion' => <<"1.0">>};
+            pv_1_1 ->
+                #{
+                    'ProtocolVersion' => <<"1.1">>,
+                    'SenderNSID' => <<"">>,
+                    'DedupWindowSize' => DedupWindowSize
+                }
+        end,
+
+    VersionBase#{
+        'SenderID' => <<"0xC00053">>,
+        'ReceiverID' => hpr_roaming_utils:hexstring(NetID),
+        'TransactionID' => TransactionID,
+        'MessageType' => <<"PRStartReq">>,
+        'PHYPayload' => hpr_roaming_utils:binary_to_hexstring(Payload),
+        'ULMetaData' => #{
+            RoutingKey => RoutingValue,
+            'DataRate' => hpr_lorawan:datarate_to_index(Region, DataRate),
+            'ULFreq' => Frequency,
+            'RecvTime' => hpr_roaming_utils:format_time(GatewayTime),
+            'RFRegion' => Region,
+            'FNSULToken' => Token,
+            'GWCnt' => erlang:length(Uplinks),
+            'GWInfo' => lists:map(fun gw_info/1, Uplinks)
+        }
+    }.
+
+%% ------------------------------------------------------------------
+%% Downlink
+%% ------------------------------------------------------------------
+
+-spec handle_message(prstart_ans() | xmitdata_req()) ->
+    ok
+    | {downlink, xmitdata_ans(), downlink(), {dest_url(), flow_type()}}
+    | {join_accept, downlink()}
+    | {error, any()}.
+handle_message(#{<<"MessageType">> := MT} = M) ->
+    case MT of
```

Comment:
Could use pattern match in function head.

Reply:
I will take a look at that.

Reply:
I don't see any significant advantage in this case.